js/services.ts

```diff
@@ -139,12 +143,127 @@ module jsonforms.services {
+            return Object.keys(properties);
+        };
+
+        private property = (instance: any) : Object => {
+            switch (typeof instance) {
+                case "string":
+                case "boolean":
+                    return { "type": typeof instance };
+                case "number":
+                    if (Number(instance) % 1 === 0) {
+                        return { "type": "integer" };
+                    } else {
+                        return { "type": "number" };
+                    }
+                case "object":
+                    return this.schemaObjectOrNullOrArray(instance);
+                default:
+                    return {};
+            }
+        };
+
+        private schemaObjectOrNullOrArray= (instance: Object): Object => {
+            if (this.isNotNull(instance)) {
+                if (this.isArray(instance)) {
+                    return this.schemaArray(<Array<Object>>instance);
+                } else {
+                    return this.schemaObject(instance);
+                }
+            } else {
+                return { "type": "null" };
+            }
+        };
+
+        private schemaArray= (instance: Array<Object>): Object => {
+            if (instance.length) {
+                var generator = this;
+                var allProperties = instance.map(function(object) {
+                    return generator.property(object);
+                });
+                var uniqueProperties = this.distinct(allProperties,
+                    function(object) { return JSON.stringify(object) });
+                if (uniqueProperties.length == 1) {
+                    return {
+                        "type": "array",
+                        "items": uniqueProperties[0]
+                    };
+                } else {
+                    return {
+                        "type": "array",
+                        "items": {
+                            "oneOf": uniqueProperties
+                        }
+                    };
+                }
+            }
+        };
+
+        private isArray = (instance: any): boolean => {
+            return Object.prototype.toString.call(instance) === '[object Array]';
+        };
+
+        private isNotNull = (instance: any): boolean => {
+            return (typeof(instance) !== 'undefined') && (instance !== null);
+        };
+
+        private distinct = (array: Array<Object>, key: (item: Object) => any): Array<Object> => {
+            var known = {};
+            return array.filter(function(item) {
+                var keyValue = key(item);
+                if (known.hasOwnProperty(keyValue)) {
+                    return false;
+                } else {
+                    return (known[keyValue] = true);
+                }
+            });
+        };
+
+        protected requiredProperties = (properties: Object): Object => {
+            return properties; // all known properties are required by default
```

Comment:
The defaults for required and additionalProperties make sense, but what if I'd like to customize this behavior? Do I need to subclass the generator or can we add additional parameters to the generateDefaultSchema method?